dimod/higherorder.py

```diff
@@ -193,28 +198,97 @@ def _reduce_degree(bqm, poly, vartype, scale):
     return _reduce_degree(bqm, new_poly, vartype, scale)
 
 
+def create_poly(linear, quadratic):
+    """ given h,j creates a single polynomial dict.
+    all h's will turn into tuples.
+
+    Args:
+        linear (dict): linear variables
+        quadratic (dict): quadratic and higher order variables
+
+    Returns
+        dict: a higher order problem dict that contains linear,
+              quadratic and n-order terms.
+
+    """
+
+    poly = {(k,): v for k, v in linear.items()}
+    poly.update(quadratic)
+    return poly
+
+
 def _prod(iterable):
     val = 1.
     for v in iterable:
         val *= v
     return val
 
 
+def _prod_d(iterable, dim):
+    val = [1.] * dim
+    for v in iterable:
+        val *= v
+    return val
+
+
 def poly_energy(sample, poly):
-    """Create a binary quadratic model from a higher order polynomial.
+    """
+    Calculate energy of a sample that is a solution to a higher order problem
+    provided by poly.
 
     Args:
-        sample (dict):
-            Sample for which to calculate the energy, formatted as a dict where keys
-            are variables and values are the value associated with each variable.
+        sample (dict or (list,np.array)):
+            Sample for which to calculate the energy, formatted as a dict
+            where keys are variables and values are the value associated with
+            each variable. If formatted as a list or np.array, the terms in
+            poly dict must be provided as variable order.
 
         poly (dict):
-            Polynomial as a dict of form {term: bias, ...}, where `term` is a tuple of
-            variables and `bias` the associated bias.
+            Polynomial as a dict of form {term: bias, ...}, where `term` is a
+            tuple of variables and `bias` the associated bias.
 
     Returns:
-        float: The energy of the sample.
+        float/list: The energy of the sample(s).
 
     """
-    return sum(_prod(sample[v] for v in variables) * bias
-               for variables, bias in poly.items())
+    if isinstance(sample, dict):
+        return poly_energies(sample, poly)
+
+    if not isinstance(sample[0], Number):
+        raise ValueError('poly_energy accepts a single sample. For multiple '
+                         'samples use poly_energies')
+    else:
+        return poly_energies(sample, poly)
+
+
+def poly_energies(samples_like, poly):
+    """
+    Calculate energy of sample(s) that are solutions to a higher order problem
+    provided by poly.
+
+    Args:
+        sample (samples_like):
+            Sample(s) for which to calculate the energy, formatted as a dict 
```

Comment:
Can probably just link to where samples-like is described elsewhere. Maybe something like
```
            samples (samples_like):
                A collection of raw samples. `samples_like` is an extension of NumPy's array_like
                structure. See :func:`.as_samples`.
```